csrc/python_frontend/translation.cpp

```diff
@@ -0,0 +1,555 @@
+      }
+    }
+
+    // Topological search of Fusion expressions
+    size_t skip_count = 0;
+    std::unordered_set<nvfuser::Expr*> visited;
+    while (!to_visit.empty()) {
+      Expr* e = to_visit.front();
+      to_visit.pop_front();
+
+      NVF_ERROR(
+          skip_count <= to_visit.size(),
+          "Cycle detected: None of the expressions can be processed!");
+
+      // short-circuit: skip if already visited
+      if (visited.count(e) > 0) {
+        continue;
+      }
+
+      // short-circuit: skip Split and Merge expressions created by Reshape
+      // short-circuit: skip Resize expressions created by Slice
+      if (e->isA<Split>() || e->isA<Merge>() || e->isA<Resize>()) {
+        visited.insert(e);
+        continue;
+      }
+
+      bool is_expr_inputs_valid =
+          std::all_of(e->inputs().begin(), e->inputs().end(), [this](Val* v) {
+            return !v->isA<TensorView>() ||
+                !isScheduledTensorView(v->as<TensorView>());
+          });
+      NVF_ERROR(
+          is_expr_inputs_valid,
+          "Found a TensorView with scheduled loop domain.");
+
+      // Handle scalars and constants not generated by separate expression.
+      std::vector<Val*> scalars;
+      std::copy_if(
+          e->inputs().begin(),
+          e->inputs().end(),
+          std::back_inserter(scalars),
+          [](Val* v) { return v->isScalar(); });
+      std::for_each(scalars.begin(), scalars.end(), [this](const Val* v) {
+        OptOutConstDispatch::dispatch(v);
+      });
+
+      // short-circuit: add to back of stack if not all of the expression's
+      // dependencies are satisfied.
+      if (!checkExpressionDependencies(e)) {
+        ++skip_count;
+        to_visit.push_back(e);
+        continue;
+      }
+
+      // Create RecordFunctor given inputs, outputs, and attributes.
+      visited.insert(e);
+      OptOutConstDispatch::dispatch(e);
+      skip_count = 0;
+    }
+
+    // Add tensor outputs and handle aliased outputs
+    std::unordered_set<nvfuser::Val*> visited_alias_output;
+    for (nvfuser::Val* v : fusion_->outputs()) {
+      NVF_ERROR(v->isA<TensorView>());
+      const AliasInfo& alias_info = fusion_->getOutputAlias(v);
+      switch (alias_info.type) {
+        case AllocationType::New: {
+          handleOutput(v->as<TensorView>());
+          break;
+        }
+        case AllocationType::ReuseBuffer: {
+          size_t num_visited = visited_alias_output.count(v);
+          if (num_visited == 0) {
+            visited_alias_output.insert(v);
+            handleOutput(v->as<TensorView>(), alias_info);
+          }
+          // An alias output can also be returned as a fusion output
+          // if it is already aliased or if not hide_output
+          if (num_visited > 0 || !alias_info.hide_output) {
+            handleOutput(v->as<TensorView>());
+          }
+          break;
+        }
+        default:
+          NVF_ERROR(false, "Unsupported AllocationType");
```

Comment:
I see Evaluate is not supported.

Don't have to add it here in this PR, I'm wondering if we can actually just ignore that.. i.e. Evaluate is supposed to be added in our transformation, since we do not have it exposed at python API where user can annotate which output gets executed by ExprEvaluator.